src/interpreter/interpreter.rs

```diff
@@ -4,22 +4,438 @@ use crate::ir::ast::Expression;
 use crate::ir::ast::Name;
 use crate::ir::ast::Statement;
 
-type IntValue = i32;
 type ErrorMessage = String;
 
-type Environment = HashMap<Name, i32>;
+#[derive(Debug, Clone, PartialEq)]
+pub enum EnvValue {
+    CInt(i32),
+    CReal(f32),
+    Bool(bool),
+    List(Vec<EvalResult>),
+    Func(
+        Box<EvalResult>,
+        Option<HashMap<Name, Box<EvalResult>>>,
+        Option<Box<Statement>>,
+        Box<Expression>,
+    ),
+    None,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum EvalResult {
+    CInt(i32),
+    CReal(f32),
+    Bool(bool),
+    List(Vec<EvalResult>),
+    None,
+}
 
-pub fn eval(exp: &Expression, env: &Environment) -> Result<IntValue, ErrorMessage> {
+type Environment = HashMap<Name, EnvValue>;
+
+pub fn eval(exp: &Expression, env: &Environment) -> Result<EvalResult, ErrorMessage> {
     match exp {
-        Expression::CInt(v) => Ok(*v),
-        Expression::Add(lhs, rhs) => Ok(eval(lhs, env)? + eval(rhs, env)?),
-        Expression::Sub(lhs, rhs) => Ok(eval(lhs, env)? - eval(rhs, env)?),
-        Expression::Mul(lhs, rhs) => Ok(eval(lhs, env)? * eval(rhs, env)?),
-        Expression::Div(lhs, rhs) => Ok(eval(lhs, env)? / eval(rhs, env)?),
+        Expression::CInt(v) => Ok(EvalResult::CInt(*v)),
+        Expression::CReal(v) => Ok(EvalResult::CReal(*v)),
+        Expression::Bool(v) => Ok(EvalResult::Bool(*v)),
+        Expression::None => Ok(EvalResult::None),
+        Expression::List(items) => {
+            let mut list_vec: Vec<EvalResult> = Vec::new();
+            let list_env = env.clone();
+
+            if items.len() < 1 {
+                return Err(String::from(
+                    "List initialization must have at least one element",
```

Comment:
Since I was not in class when this decision was possibly made, I have to ask: was it explicitly decided that we can't initialise an empty list?